### PR TITLE
Prevent BatchEncoding from blindly passing casts down to the tensors it contains

### LINKED ISSUE #6582: BatchEncoding interacts poorly with apex.amp
## Environment info
- `transformers` version: 3.0.0
- Platform: Linux
- Python version: 3.6
- PyTorch version (GPU?): 1.6.0 GPU

### Who can help
tokenizers: @mfuntowicz 

## Information

We are building a multi-task transformer on top of the transformers API, using Nvidia's Apex to get the best speed out of the GPUs. When using Apex with `O2` optimisations (which are necessary to get speedups on our model), it inserts casts around the inputs to `forward` which in our case is a `BatchEncoding`. This cast blindly calls `BatchEncoding.to` with a `dtype` argument, which in turn is passed through to the values `to` method because there is only a type hint `str` on the `BatchEncoding.to` input argument, not a `isinstance` check. This causes our tensor of BPE ids to be cast from a `LongTensor` into `HalfTensor` and then the embedding layer raises an error because it's not expecting floating point values (also we rounded away most of our vocabulary terms).

## To reproduce

Run amp.initialize with the `O2` opt level on a model which's forward method accepts BatchEncoding as the input.

This [line](https://github.com/NVIDIA/apex/blob/4ef930c1c884fdca5f472ab2ce7cb9b505d26c1a/apex/amp/_initialize.py#L46) in apex calls the `to` method blindly on all the input types, which in this case calls `BatchEncoding.to` which doesn't support the full scope of a pytorch `to` because it's designed for moving things between devices and not for casting them.

We're currently guarding it this way:
```python
def patched_to(self, device: Union[str,torch.device]):
    """Send all values to device by calling v.to(device)"""
    if isinstance(device, str) or isinstance(device, torch.device):
        self.data = {k: v.to(device=device) for k, v in self.data.items()}
    return self
BatchEncoding.to = patched_to
```

## Expected behavior

The `BatchEncoding.to` method to not to pass the cast through (as the documentation indicates).

Though it's not clear if this is because the `to` method is underspecified, or because Apex is being quite aggressive (and a little silly) in it's blind casting of everything. Either way, if the `BatchEncoding.to` method isn't supposed to be able to cast things (and it doesn't look like it) then it probably shouldn't pass casts down to the tensors it contains.

### PR DESCRIPTION
**Pull request by Craigacp**
# What does this PR do?

This PR prevents `BatchEncoding.to` from passing down things which aren't devices to the tensors it contains. Previously it would pass down all the arguments, and as the `to` method in pytorch can also cast the arguments to different types it's used blindly by other packages (e.g. Nvidia's Apex). This caused an issue where when using Apex's AMP support with `O2` or greater it would cast the token indexes from a `LongTensor` to a `HalfTensor` truncating our vocab at 65k and rounding most of the words to the nearest 8th word (if you blindly insert the cast back in in the embedding layer, which the warning says to do).

The doc for `BatchEncoding.to` says it is only for moving the encoding and the tensors it contains between devices, but as the type checking isn't on by default it can behave like a regular pytorch `to` method and accept cast arguments that it passes down to the tensors it contains.

Fixes #6582

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. #6582
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

There are no docs or tests changes as the change makes the method conform with its currently documented behaviour.

@LysandreJik

### CODE REVIEW

File: src/transformers/tokenization_utils_base.py [hunk 1]
```diff
@@ -776,7 +776,14 @@ def to(self, device: Union[str, "torch.device"]) -> "BatchEncoding":
...
         """
-        self.data = {k: v.to(device) for k, v in self.data.items()}
+
+        # This check catches things like APEX blindly calling "to" on all inputs to a module
+        # Otherwise it passes the casts down and casts the LongTensor containing the token idxs
+        # into a HalfTensor
+        if isinstance(device, str) or isinstance(device, torch.device):
+            self.data = {k: v.to(device=device) for k, v in self.data.items()}
+        else:
+            logger.warning("Attempting to cast a BatchEncoding to another type, %s", str(device))
```
Comment (LysandreJik) [inline src/transformers/tokenization_utils_base.py:786]:
```suggestion
            logger.warning(f"Attempting to cast a BatchEncoding to another type, {str(device)}. This is not supported.")
```

**Review by LysandreJik — APPROVED** [review]
LGTM! We prefer using f-strings so I suggested a change. Thank you for opening the PR.

File: src/transformers/tokenization_utils_base.py (hunk 1 above)
Comment (Craigacp) [reply src/transformers/tokenization_utils_base.py]: Sure, fine by me. I think I copied a logger statement from further down the file, so there are some others outside of this commit that need fixing in that file.

**Comment by Craigacp** [comment]
black complained about the style after the update, so I fixed it and squashed the commits again.

**Comment by LysandreJik** [comment]
Thank you @Craigacp!

**Pull request merged**